### ISSUE TITLE
Update XML definition

### ISSUE THREAD

**Issue by WJSchakel** [averbraeck/opentrafficsim#40]
Let's first collect several small issues, so we can do this in one batch. You can edit this message.

- CONNECTOR tag needs a TYPE (link type) so GTU compatibility can be defined on it. Connector constructor also needs it.<sup>1</sup>
- default_gtutypes.xml: remove NONE, name all others NL.XXX.<sup>1</sup>
- default_linktypes.xml: remove NONE, name all others NL.XXX.<sup>1</sup>
- default_lanetypes.xml: remove NONE, name all others NL.XXX.<sup>1</sup>
- What to do with DIRECTION in COMPATIBILITY of both LINKTYPE and LANETYPE?

<sup>1</sup>) Needs a change in the XML parser.

**Comment by WJSchakel** [comment]
In XSD we have a great number of restrictions that are defined with patterns. In case of a pure choice:

```xml
  <xsd:simpleType name="LANEKEEPINGTYPE">
    <xsd:restriction base="xsd:string">
      <xsd:pattern value="KEEPLEFT|KEEPRIGHT|KEEPLANE"></xsd:pattern>
    </xsd:restriction>
  </xsd:simpleType>
```

It would help to defined this as an enumeration:

```xml
  <xsd:simpleType name="LANEKEEPINGTYPE">
    <xsd:restriction base="xsd:string">
      <xsd:enumeration value="KEEPLEFT" />
      <xsd:enumeration value="KEEPRIGHT" />
      <xsd:enumeration value="KEEPLANE" />
    </xsd:restriction>
  </xsd:simpleType>
```

In this way, we can present a dropdown-menu/pick-list to users.

**Comment by averbraeck** [comment]
Please wait one day till I addresses Issue #55. The XSD files are being changed for that issue at the moment.

**Comment by WJSchakel** [comment]
Yeah sure. I think everything in this issue should be done in one go at some later point.

**Comment by OTSim** [comment]
This is also a test to see if replies go to the right place…

I would go for the annotation approach. This ensures that you will be aware that you are about to break things if you want to change that part of the xsd. Of course, it does not prevent you from breaking things, but all the structure and interpretation of this part of the xsd is all in one place (unlike having the interpretation embedded in the code of the ots-editor).
--
PK

**Comment by WJSchakel** [comment]
Yes, reply works! It does come with the text that you replay to, though. That clutters things so I'd suggest removing that text.

Right now, the code looks for the annotation. If it does not find it, it constructs a name from the sequence. With a max length of 64 characters.

**Comment by WJSchakel** [comment]
Centroid is added to XML specification. Checks on Connectors towards Node id's have been removed, as Connectors can connect to Centroids too.

**Comment by WJSchakel** [comment]
`DetectorType` tag fully integrated in parsing. "default_detectortypes.xml" added to all XML files.

**Comment by WJSchakel** [comment]
Gtu Colorers are now defined in one of 3 ways, through an `xsd:choice`:

- A default options (ID, SPEED, ACCELERATION, etc.)
- An incentive option (ROUTE, SPEED, COURTESY, etc.)
- A class

Animation is not yet parsed, and we may add more.

**Comment by WJSchakel** [comment]
When a value does not match a pattern, a sensible message can now be defined as:

```
  <xsd:simpleType name="LengthType">
    <xsd:annotation>
      <xsd:appinfo source="pattern">Provide a valid Length, such as "3.5m".</xsd:appinfo>
    </xsd:annotation>
    <xsd:restriction base="xsd:string">
      <xsd:pattern value="[+-]?[0-9]*\.?[0-9]+([eE]......"></xsd:pattern>
    </xsd:restriction>
  </xsd:simpleType>
```

**Comment by WJSchakel** [comment]
Random streams can be defined, or defaults ("default" or "generation") can be used. To be able to check whether a selected random stream is defined, an `xsd:keyref` is defined. The default streams can be used by leaving the option blank, in which case the default for the particular type ("default" or "generation") will be used.

All constant and discrete distribution types stem from a single `DistType` with `RandomStream` attribute. An `xsd:keyref` to this attribute is defined. The editor now also deal with keyref's in extended types.

**Comment by WJSchakel** [comment]
RandomStream definitions has been enhanced in XML as per issue #47.

**Comment by WJSchakel** [comment]
There is now a list `Ots.Scenarios.DefaultInputParameters`. Two things may point to this:

1) Values that are expressions that change between scenarios, using { }.
2) `Ots.Scenarios.Scenario.InputParameters` to overwrite default values for a specific scenario.
